[4.4] Rebuild mod_users_latest with new Joomla 4 structure

sandewt opened the pull request:
Pull Request for Issue # .

### Summary of Changes

Rebuild mod_users_latest with new Joomla 4 structure

### Testing Instructions

- Code review
- Just try to use it as usual

### Actual result BEFORE applying this Pull Request

- Module works

### Expected result AFTER applying this Pull Request

- Module works

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

sandewt committed:
Update mod_users_latest.xml

sandewt committed:
Update UsersLatestHelper.php

sandewt committed:
Create Dispatcher.php

sandewt committed:
Delete modules/mod_users_latest/Dispatcher directory

sandewt committed:
Create provider.php

sandewt committed:
Create Dispatcher.php

sandewt committed:
Delete mod_users_latest.php

sandewt commented:
Related to:

#39931 : mod_custom
#40066 : mod_footer

sandewt committed:
Delete Factory class

Delete Factory class

carlitorweb commented on modules/mod_users_latest/mod_users_latest.xml:
```diff
@@ -11,7 +11,7 @@
 	<namespace path="src">Joomla\Module\UsersLatest</namespace>
 	<files>
-		<filename module="mod_users_latest">mod_users_latest.php</filename>
+		<folder module="mod_users_latest">mod_users_latest.php</folder>
```
`<folder module="mod_users_latest">src</folder>`

carlitorweb commented on modules/mod_users_latest/mod_users_latest.xml:
```diff
@@ -11,7 +11,7 @@
 	<namespace path="src">Joomla\Module\UsersLatest</namespace>
 	<files>
-		<filename module="mod_users_latest">mod_users_latest.php</filename>
+		<folder module="mod_users_latest">mod_users_latest.php</folder>
 		<folder>src</folder>
 		<folder>tmpl</folder>
```
```xml
<folder module="mod_users_latest">src</folder>
<folder>services</folder>
<folder>tmpl</folder>
```

sandewt replied:
Thanks, I was mistaken.

The next variant is generally used, see eg `mod_articles_latest.xml` and so on

```
<folder module="mod_articles_latest">services</folder>
<folder>src</folder>
<folder>tmpl</folder>
```

And with the plugins, it is used, eg:
```
<folder plugin="cookie">services</folder>
<folder>src</folder>
```

So I think this is most desirable:
`<folder module="mod_users_latest">services</folder>`

sandewt replied:
@laoneo  I think a separate PR for the `modules` and `plugins`  is desirable to make this code consistent throughout.

carlitorweb replied:
@sandewt  is services for the module, you are correct, apologies

sandewt replied:
@carlitorweb Your proposal was quite correct. The code of the `mod_articles_news` module has the same structure as you suggested. That's why I suggest a separate PR to make the code more consistent.

laoneo replied:
Agree that a separate pr would be a good idea.

sandewt replied:
> Agree that a separate pr would be a good idea.

I'll do that.

sandewt replied:
> Agree that a separate pr would be a good idea.

@laoneo pr for J4.3 or J4.4 ?

laoneo replied:
Better 4.4.

sandewt replied:
See pr #40204

sandewt committed:
services

`folder module="mod_users_latest">mod_users_latest.php</folder>`
becomes
`<folder module="mod_articles_latest">services</folder>`

laoneo commented:
I worked today on system tests and I'm on a point now where it is relative easy to write one for front end modules. It would be cool if you can also write one for the converted module here. It is not that much of code to write. An example can be found here https://github.com/joomla/joomla-cms/pull/40102 (ignore the changes in the db file).

laoneo committed:
Merge branch '4.4-dev' into patch-4

sandewt commented:
> It would be cool if you can also write one for the converted module here. It is not that much of code to write. 

Sounds interesting. I'm going to check it out. But don't promise anything yet. Come back to it.

toivo commented:
I have tested this item :white_check_mark: successfully on d3ad059db0dd9497410ecef3a299f4ceabac8090<br /><br />Tested successfully in Joomla 4.4.0-dev of 14 March using PHP 8.1.10.

The following Deprecated message was reported from the back end when creating the front tne module, with or without the PR:
```
[14-Mar-2023 02:09:13 UTC] PHP Deprecated:  strlen(): Passing null to parameter #1 ($string) of type string is deprecated in C:\www\joomla44\libraries\src\Table\Module.php on line 152
[14-Mar-2023 02:09:13 UTC] PHP Stack trace:
[14-Mar-2023 02:09:13 UTC] PHP   1. {main}() C:\www\joomla44\administrator\index.php:0
[14-Mar-2023 02:09:13 UTC] PHP   2. require_once() C:\www\joomla44\administrator\index.php:32
[14-Mar-2023 02:09:13 UTC] PHP   3. Joomla\CMS\Application\CMSApplication->execute() C:\www\joomla44\administrator\includes\app.php:61
[14-Mar-2023 02:09:13 UTC] PHP   4. Joomla\CMS\Application\AdministratorApplication->doExecute() C:\www\joomla44\libraries\src\Application\CMSApplication.php:293
[14-Mar-2023 02:09:13 UTC] PHP   5. Joomla\CMS\Application\AdministratorApplication->dispatch($component = *uninitialized*) C:\www\joomla44\libraries\src\Application\AdministratorApplication.php:186
[14-Mar-2023 02:09:13 UTC] PHP   6. Joomla\CMS\Component\ComponentHelper::renderComponent($option = 'com_modules', $params = *uninitialized*) C:\www\joomla44\libraries\src\Application\AdministratorApplication.php:143
[14-Mar-2023 02:09:13 UTC] PHP   7. Joomla\CMS\Dispatcher\ComponentDispatcher->dispatch() C:\www\joomla44\libraries\src\Component\ComponentHelper.php:355
[14-Mar-2023 02:09:13 UTC] PHP   8. Joomla\CMS\MVC\Controller\BaseController->execute($task = 'save') C:\www\joomla44\libraries\src\Dispatcher\ComponentDispatcher.php:143
[14-Mar-2023 02:09:13 UTC] PHP   9. Joomla\Component\Modules\Administrator\Controller\ModuleController->save($key = *uninitialized*, $urlVar = *uninitialized*) C:\www\joomla44\libraries\src\MVC\Controller\BaseController.php:672
[14-Mar-2023 02:09:13 UTC] PHP  10. Joomla\CMS\MVC\Controller\FormController->save($key = NULL, $urlVar = NULL) C:\www\joomla44\administrator\components\com_modules\src\Controller\ModuleController.php:241
[14-Mar-2023 02:09:13 UTC] PHP  11. Joomla\Component\Modules\Administrator\Model\ModuleModel->save($data = ['id' => '0', 'title' => 'Latest Users 2', 'note' => '', 'module' => 'mod_users_latest', 'showtitle' => '1', 'published' => '1', 'publish_up' => '', 'publish_down' => '', 'client_id' => '0', 'position' => 'sidebar-right', 'access' => 1, 'ordering' => '1', 'language' => '*', 'assignment' => '0', 'assigned' => [0 => '101'], 'rules' => ['core.delete' => [...], 'core.edit' => [...], 'core.edit.state' => [...], 'module.edit.frontend' => [...]], 'params' => ['shownumber' => 5, 'filter_groups' => 0, 'layout' => '_:default', 'moduleclass_sfx' => '', 'cache' => 1, 'cache_time' => 900, 'cachemode' => 'static', 'module_tag' => 'div', 'bootstrap_size' => '0', 'header_tag' => 'h3', 'header_class' => '', 'style' => '0'], 'tags' => []]) C:\www\joomla44\libraries\src\MVC\Controller\FormController.php:675
[14-Mar-2023 02:09:13 UTC] PHP  12. Joomla\CMS\Table\Module->check() C:\www\joomla44\administrator\components\com_modules\src\Model\ModuleModel.php:934
```<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/40092">issues.joomla.org/tracker/joomla-cms/40092</a>.</sub>

sandewt committed:
cleaner code

replace  $user = $app->getIdentity();

sandewt commented:
> strlen(): Passing null to parameter #1 ($string) of type string is deprecated in C:\www\joomla44\libraries\src\Table\Module.php on line 152

As far as I can see this is a `PHP 8.1 error: strlen()` and is unrelated to this pr. I think a separate PR is needed to solve this. Also because this error occurs before and after this pr.

sandewt committed:
Add mod_users_latest system test

mod_users_latest system test

laoneo commented:
You have a syntax code style error in the test https://ci.joomla.org/joomla/joomla-cms/63196/1/29.

sandewt committed:
correction

syntax error

laoneo commented:
You need to create a user in your test, add the following code to the file tests/cypress/support/commands/db.js after line 61:
```
Cypress.Commands.add('db_createUser', (user) => {
  const defaultUserOptions = {
    name: 'test user',
    username: 'test',
    email: 'test@example.com',
    password: '098f6bcd4621d373cade4e832627b4f6', // Is the md5 of the word 'test'
    block: 0
  };
  user = { ...defaultUserOptions, ...user };

  const groupId = user.group_id ?? 2; // Default the group id to registered
  delete user.group_id;

  return cy.task('queryDB', createInsertQuery('users', user)).then((info) => {
    cy.task('queryDB', "INSERT INTO #__user_usergroup_map (user_id, group_id) VALUES ('" + info.insertId + "', '" + groupId + "')").then(() => info);
  });
});
```

And then replace your test with:
```
describe('Test that the users latest module', () => {
  it('can load in frontend and displays the latest registered users', () => {
    cy.db_createUser({ username: 'automatedtestuser' })
      .then(() => cy.db_createModule({ module: 'mod_users_latest' }))
      .then(() => {
        cy.visit('/');

        cy.contains('automatedtestuser');
      });
  });
});
```

laoneo committed:
Merge branch '4.4-dev' into patch-4

sandewt committed:
replace code

sandewt committed:
create a user

create a user for the test

sandewt committed:
add registerDate

registerDate: '2023-03-01 20:00:00'

sandewt committed:
add params

params: ''

sandewt commented on tests/cypress/support/commands/db.js:
```diff
@@ -60,6 +60,26 @@ Cypress.Commands.add('db_createModule', (module) => {
+    email: 'test@example.com',
+    password: '098f6bcd4621d373cade4e832627b4f6', // Is the md5 of the word 'test'
+    block: 0,
+    registerDate: '2023-03-01 20:00:00',
+    params: ''
```
@laoneo 
I had to add these two parameters.
For the date I used a more or less random date. Do you know a nicer way to be consistent?

laoneo replied:
You can use the Date object and build it from the current time as described here https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date

sandewt replied:
This could possibly be the start of a solution. Actually, I want to leave it at that.

```
var date = new Date();
var date = date.toISOString().match(/(\d{4}\-\d{2}\-\d{2})T(\d{2}:\d{2}:\d{2})/)

console.log(date); // Array ["2023-03-15T16:18:27", "2023-03-15", "16:18:27"]
console.log(date[1] + ' ' + date[2]); // "2023-03-15 16:18:27"
```

laoneo commented:
Thanks!

joomdonation commented:
@laoneo @sandewt On other modules which we converted so far, we usually keep the static method in Helper class for backward compatible purpose, like this one https://github.com/joomla/joomla-cms/blob/4.2-dev/modules/mod_articles_latest/src/Helper/ArticlesLatestHelper.php#L161-L164

Look like you forgot that in this OR ?

sandewt commented:
> Look like you forgot that in this OR ?

Does this mean this module will not work in J4.2 and J4.3 ?

laoneo commented:
@joomdonation is right, can somebody make a followup pr?



> Does this mean this module will not work in J4.2 and J4.3 ?

It means when we ship this code it can break other extensions or templates.

sandewt commented:
@joomdonation Glad you discovered this error. 👍 A teaching moment for me.

laoneo commented:
@sandewt can you make sure that the other module pr's will also be backwards compatible?

sandewt commented:
> @sandewt can you make sure that the other module pr's will also be backwards compatible?

Good question.

`mod_footer` #40066 has/had no static method (in Helper class).  This is the other module I modified.
